File: cpp/src/writer.cpp
#include "zerobuffer/writer.h"
#include <cstring>
#include <atomic>

namespace zerobuffer {

class Writer::Impl {
public:
    explicit Impl(const std::string& name) 
        : name_(name), sequence_number_(1), frames_written_(0), bytes_written_(0), 
          metadata_written_(false), pending_metadata_size_(0), pending_frame_size_(0),
          pending_frame_sequence_(0), pending_frame_total_size_(0), pending_write_pos_(0) {
        
        // Open existing shared memory
        shm_ = SharedMemory::open(name);
        
        // Verify OIEB
        OIEB* oieb = get_oieb();
        if (oieb->operation_size != sizeof(OIEB)) {
            throw ZeroBufferException("Invalid OIEB size - version mismatch?");
        }
        
        // Check if reader exists
        if (oieb->reader_pid == 0 || !platform::process_exists(oieb->reader_pid)) {
            throw ZeroBufferException("No active reader found");
        }
        
        // Check if another writer exists
        if (oieb->writer_pid != 0 && platform::process_exists(oieb->writer_pid)) {
            throw ZeroBufferException("Another writer is already connected");
        }
        
        // Set writer PID
        oieb->writer_pid = platform::get_current_pid();
        
        // Open semaphores
        sem_write_ = Semaphore::open("/sem-w-" + name);
        sem_read_ = Semaphore::open("/sem-r-" + name);
        
        // Set up pointers
        metadata_start_ = reinterpret_cast<uint8_t*>(shm_->data()) + sizeof(OIEB);
        payload_start_ = metadata_start_ + oieb->metadata_size;
        
        // Check if metadata was already written
        metadata_written_ = (oieb->metadata_written_bytes > 0);
    }
    
    ~Impl() {
        // Clear writer PID
        if (shm_ && shm_->valid()) {
            OIEB* oieb = get_oieb();
            oieb->writer_pid = 0;
        }
    }
    
    OIEB* get_oieb() {
        return reinterpret_cast<OIEB*>(shm_->data());
    }
    
    const OIEB* get_oieb() const {
        return reinterpret_cast<const OIEB*>(shm_->data());
    }
    
    void* get_metadata_buffer(size_t size) {
        if (metadata_written_) {
            throw MetadataAlreadyWrittenException();
        }
        
        OIEB* oieb = get_oieb();
        
        // Check size with header
        size_t total_size = sizeof(uint64_t) + size;
        if (total_size > oieb->metadata_size) {
            throw ZeroBufferException("Metadata too large for buffer");
        }
        
        // Store pending size
        pending_metadata_size_ = size;
        
        // Return pointer to data area (after size prefix)
        return metadata_start_ + sizeof(uint64_t);
    }
    
    void commit_metadata() {
        if (metadata_written_) {
            throw MetadataAlreadyWrittenException();
        }
        
        OIEB* oieb = get_oieb();
        
        // Write size prefix
        uint64_t meta_size = pending_metadata_size_;
        std::memcpy(metadata_start_, &meta_size, sizeof(meta_size));
        
        // Update OIEB
        size_t total_size = sizeof(uint64_t) + pending_metadata_size_;
        oieb->metadata_written_bytes = total_size;
        oieb->metadata_free_bytes = oieb->metadata_size - total_size;
        
        metadata_written_ = true;
    }
    
    void set_metadata(const void* data, size_t size) {
        if (metadata_written_) {
            throw MetadataAlreadyWrittenException();
        }
        
        OIEB* oieb = get_oieb();
        
        // Check size with header
        size_t total_size = sizeof(uint64_t) + size;
        if (total_size > oieb->metadata_size) {
            throw ZeroBufferException("Metadata too large for buffer");
        }
        
        // Write size prefix
        uint64_t meta_size = size;
        std::memcpy(metadata_start_, &meta_size, sizeof(meta_size));
        
        // Write metadata
        if (size > 0) {
            std::memcpy(metadata_start_ + sizeof(meta_size), data, size);
        }
        
        // Update OIEB
        oieb->metadata_written_bytes = total_size;
        oieb->metadata_free_bytes = oieb->metadata_size - total_size;
        
        metadata_written_ = true;
    }
    
    size_t get_continuous_free_space() const {
        const OIEB* oieb = get_oieb();
        
        // Calculate continuous free space based on read/write positions
        if (oieb->payload_write_pos >= oieb->payload_read_pos) {
            // Write ahead of read - check space to end and beginning
            size_t space_to_end = oieb->payload_size - oieb->payload_write_pos;
            if (oieb->payload_read_pos == 0) {
                // Can't wrap if reader at beginning
                return space_to_end;
            }
            // Can use space at beginning if we wrap
            return std::max(space_to_end, oieb->payload_read_pos);
        } else {
            // Read ahead of write - continuous space until read pos
            return oieb->payload_read_pos - oieb->payload_write_pos;
        }
    }
    
    void write_frame(const void* data, size_t size) {
        if (size == 0) {
            throw InvalidFrameSizeException();
        }
        
        size_t total_size = sizeof(FrameHeader) + size;
        
        while (true) {
            OIEB* oieb = get_oieb();
            
            // Check if reader is still alive
            if (!is_reader_connected()) {
                throw ReaderDeadException();
            }
            
            // Check continuous free space
            size_t continuous_free = get_continuous_free_space();
            
            if (continuous_free >= total_size) {
                // We have enough space
                break;
            }
            
            // Wait for reader to free space
            if (!sem_read_->wait(std::chrono::milliseconds(5000))) {
                // Timeout - check if reader is alive
                if (!is_reader_connected()) {
                    throw ReaderDeadException();
                }
                // Continue waiting
            }
        }
        
        OIEB* oieb = get_oieb();
        
        // Check if we need to wrap
        size_t space_to_end = oieb->payload_size - oieb->payload_write_pos;
        if (space_to_end < total_size && oieb->payload_read_pos > 0) {
            // Need to wrap to beginning
            // Write a special marker at current position if there's space for a header
            if (space_to_end >= sizeof(FrameHeader)) {
                uint8_t* wrap_marker_ptr = payload_start_ + oieb->payload_write_pos;
                // Write wrap marker header
                FrameHeader wrap_marker;
                wrap_marker.payload_size = 0;  // Indicates wrap-around
                wrap_marker.sequence_number = 0;
                std::memcpy(wrap_marker_ptr, &wrap_marker, sizeof(wrap_marker));
                
                // Increment written count for the wrap marker
                oieb->payload_written_count++;
            }
            // Account for the wasted space at the end
            oieb->payload_free_bytes -= space_to_end;
            oieb->payload_write_pos = 0;
        }
        
        // Write frame header
        uint8_t* write_ptr = payload_start_ + oieb->payload_write_pos;
        FrameHeader header;
        header.payload_size = size;
        header.sequence_number = sequence_number_;
        std::memcpy(write_ptr, &header, sizeof(header));
        
        // Write frame data
        std::memcpy(write_ptr + sizeof(header), data, size);
        
        // Update tracking
        oieb->payload_write_pos += total_size;
        sequence_number_++;
        frames_written_++;
        bytes_written_ += size;
        
        // Update OIEB
        oieb->payload_free_bytes -= total_size;
        oieb->payload_written_count++;
        
        // Release memory barrier
        std::atomic_thread_fence(std::memory_order_release);
        
        // Signal reader
        sem_write_->signal();
    }
    
    void* get_frame_buffer(size_t size, uint64_t& sequence_number) {
        if (size == 0) {
            throw InvalidFrameSizeException();
        }
        
        size_t total_size = sizeof(FrameHeader) + size;
        
        while (true) {
            OIEB* oieb = get_oieb();
            
            // Check if reader is still alive
            if (!is_reader_connected()) {
                throw ReaderDeadException();
            }
            
            // Check continuous free space
            size_t continuous_free = get_continuous_free_space();
            
            if (continuous_free >= total_size) {
                // We have enough space
                break;
            }
            
            // Wait for reader to free space
            if (!sem_read_->wait(std::chrono::milliseconds(5000))) {
                // Timeout - check if reader is alive
                if (!is_reader_connected()) {
                    throw ReaderDeadException();
                }
                // Continue waiting
            }
        }
        
        OIEB* oieb = get_oieb();
        
        // Check if we need to wrap
        size_t space_to_end = oieb->payload_size - oieb->payload_write_pos;
        if (space_to_end < total_size && oieb->payload_read_pos > 0) {
            // Need to wrap to beginning
            // Write a special marker at current position if there's space for a header
            if (space_to_end >= sizeof(FrameHeader)) {
                uint8_t* wrap_marker_ptr = payload_start_ + oieb->payload_write_pos;
                // Write wrap marker header
                FrameHeader wrap_marker;
                wrap_marker.payload_size = 0;  // Indicates wrap-around
                wrap_marker.sequence_number = 0;
                std::memcpy(wrap_marker_ptr, &wrap_marker, sizeof(wrap_marker));
                
                // Increment written count for the wrap marker
                oieb->payload_written_count++;
            }
            // Account for the wasted space at the end
            oieb->payload_free_bytes -= space_to_end;
            oieb->payload_write_pos = 0;
        }
        
        // Store pending operation info
        pending_write_pos_ = oieb->payload_write_pos;
        pending_frame_size_ = size;
        pending_frame_total_size_ = total_size;
        sequence_number = pending_frame_sequence_ = sequence_number_++;
        
        // Write frame header
        uint8_t* write_ptr = payload_start_ + oieb->payload_write_pos;
        FrameHeader header;
        header.payload_size = size;
        header.sequence_number = pending_frame_sequence_;
        std::memcpy(write_ptr, &header, sizeof(header));
        
        // Return pointer to frame data area
        return write_ptr + sizeof(header);
    }
    
    void commit_frame() {
        OIEB* oieb = get_oieb();
        
        // Update tracking
        oieb->payload_write_pos += pending_frame_total_size_;
        frames_written_++;
        bytes_written_ += pending_frame_size_;
        
        // Update OIEB
        oieb->payload_free_bytes -= pending_frame_total_size_;
        oieb->payload_written_count++;
        
        // Release memory barrier
        std::atomic_thread_fence(std::memory_order_release);
        
        // Signal reader
        sem_write_->signal();
    }
    
    bool is_reader_connected() const {
        const OIEB* oieb = get_oieb();
        return oieb->reader_pid != 0 && platform::process_exists(oieb->reader_pid);
    }
    
private:
    std::string name_;
    std::unique_ptr<SharedMemory> shm_;
    std::unique_ptr<Semaphore> sem_write_;
    std::unique_ptr<Semaphore> sem_read_;
    
    uint8_t* metadata_start_;
    uint8_t* payload_start_;
    uint64_t sequence_number_;
    uint64_t frames_written_;
    uint64_t bytes_written_;
    bool metadata_written_;
    
    // For zero-copy operations
    size_t pending_metadata_size_;
    size_t pending_frame_size_;
    uint64_t pending_frame_sequence_;
    size_t pending_frame_total_size_;
    size_t pending_write_pos_;
    
    friend class Writer;  // Allow Writer to access private members
};

Writer::Writer(const std::string& name)
    : impl_(std::make_unique<Impl>(name)), name_(name) {}

Writer::~Writer() = default;

Writer::Writer(Writer&&) noexcept = default;
Writer& Writer::operator=(Writer&&) noexcept = default;

void Writer::set_metadata(const void* data, size_t size) {
    impl_->set_metadata(data, size);
}

void Writer::set_metadata(const std::vector<uint8_t>& data) {
    impl_->set_metadata(data.data(), data.size());
}

void Writer::write_frame(const void* data, size_t size) {
    impl_->write_frame(data, size);
}

void Writer::write_frame(const std::vector<uint8_t>& data) {
    impl_->write_frame(data.data(), data.size());
}

void* Writer::get_metadata_buffer(size_t size) {
    return impl_->get_metadata_buffer(size);
}

void Writer::commit_metadata() {
    impl_->commit_metadata();
}

void* Writer::get_frame_buffer(size_t size, uint64_t& sequence_number) {
    return impl_->get_frame_buffer(size, sequence_number);
}

void Writer::commit_frame() {
    impl_->commit_frame();
}

bool Writer::is_reader_connected() const {
    return impl_->is_reader_connected();
}

uint64_t Writer::frames_written() const {
    return impl_->frames_written_;
}

uint64_t Writer::bytes_written() const {
    return impl_->bytes_written_;
}

} // namespace zerobuffer

File: csharp/ZeroBuffer/Writer.cs
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace ZeroBuffer
{
    /// <summary>
    /// Cross-platform Writer implementation using abstracted shared memory and semaphores
    /// </summary>
    public sealed class Writer : IDisposable
    {
        private readonly string _name;
        private readonly ILogger<Writer> _logger;
        private readonly ISharedMemory _sharedMemory;
        private readonly ISemaphore _writeSemaphore;
        private readonly ISemaphore _readSemaphore;
        private readonly long _metadataOffset;
        private readonly long _payloadOffset;
        private bool _disposed;
        private bool _metadataWritten;
        private ulong _sequenceNumber = 1;
        private TimeSpan _writeTimeout = TimeSpan.FromSeconds(5);

        /// <summary>
        /// Gets or sets the timeout for write operations when the buffer is full.
        /// Default is 5 seconds.
        /// </summary>
        public TimeSpan WriteTimeout 
        { 
            get => _writeTimeout;
            set 
            {
                if (value <= TimeSpan.Zero)
                    throw new ArgumentOutOfRangeException(nameof(value), "Write timeout must be positive");
                _writeTimeout = value;
            }
        }

        public Writer(string name) : this(name, NullLogger<Writer>.Instance)
        {
        }

        public Writer(string name, ILogger<Writer> logger)
        {
            ArgumentException.ThrowIfNullOrEmpty(name);
            ArgumentNullException.ThrowIfNull(logger);

            _name = name;
            _logger = logger;

            try
            {
                // Open shared memory first to check if buffer exists
                _sharedMemory = SharedMemoryFactory.Open(name);
                
                // Only try to open semaphores if shared memory exists
                _writeSemaphore = SemaphoreFactory.Open($"sem-w-{name}");
                _readSemaphore = SemaphoreFactory.Open($"sem-r-{name}");

                // Read OIEB to get layout - use ref for initial read-only access
                ref readonly var oiebRef = ref _sharedMemory.ReadRef<OIEB>(0);
                
                // Verify OIEB size and version compatibility
                if (oiebRef.OiebSize != 128)
                {
                    throw new InvalidOperationException("Invalid OIEB size - expected 128 for v1.x.x");
                }
                
                var currentVersion = new ProtocolVersion(1, 0, 0);
                if (!oiebRef.Version.IsCompatibleWith(currentVersion))
                {
                    throw new InvalidOperationException("Incompatible protocol version");
                }
                
                // Calculate offsets (OIEB is always 128 bytes)
                _metadataOffset = 128;  // sizeof(OIEB)
                _payloadOffset = _metadataOffset + (long)oiebRef.MetadataSize;

                // Check if another writer is already connected
                if (oiebRef.WriterPid != 0 && ProcessExists(oiebRef.WriterPid))
                {
                    throw new WriterAlreadyConnectedException();
                }

                // Get a reference to set our PID directly
                ref var oieb = ref _sharedMemory.ReadRef<OIEB>(0);
                oieb.WriterPid = (ulong)Environment.ProcessId;
                _sharedMemory.Flush();
            }
            catch
            {
                Dispose();
                throw;
            }
        }

        /// <summary>
        /// Get metadata buffer for zero-copy writing
        /// Returns a span where you can write metadata directly
        /// </summary>
        public unsafe Span<byte> GetMetadataBuffer(int size)
        {
            ThrowIfDisposed();

            if (_metadataWritten)
                throw new InvalidOperationException("Metadata already written");

            ref readonly var oieb = ref _sharedMemory.ReadRef<OIEB>(0);

            // Total size includes 8-byte size prefix
            ulong totalSize = (ulong)size + sizeof(ulong);
            if (totalSize > oieb.MetadataSize)
                throw new ArgumentException("Metadata too large for buffer");

            // Get pointer to metadata area
            byte* metadataPtr = _sharedMemory.GetPointer(_metadataOffset);
            
            // Write size prefix (8 bytes)
            *(ulong*)metadataPtr = (ulong)size;
            
            // Store size for commit
            _pendingMetadataSize = (int)totalSize;
            
            // Return buffer starting after the size prefix
            return new Span<byte>(metadataPtr + sizeof(ulong), size);
        }
        
        private int _pendingMetadataSize;
        
        /// <summary>
        /// Commit metadata after writing to the buffer returned by GetMetadataBuffer
        /// </summary>
        public void CommitMetadata()
        {
            ref var oieb = ref _sharedMemory.ReadRef<OIEB>(0);

            // Update OIEB with total size (including 8-byte prefix)
            oieb.MetadataWrittenBytes = (ulong)_pendingMetadataSize;
            oieb.MetadataFreeBytes = oieb.MetadataSize - (ulong)_pendingMetadataSize;
            _sharedMemory.Flush();

            _metadataWritten = true;
        }

        /// <summary>
        /// Set metadata (convenience method that copies data)
        /// For zero-copy writing, use GetMetadataBuffer/CommitMetadata instead
        /// </summary>
        public void SetMetadata(ReadOnlySpan<byte> data)
        {
            var buffer = GetMetadataBuffer(data.Length);
            data.CopyTo(buffer);
            CommitMetadata();
        }

        /// <summary>
        /// Get a buffer for writing frame data directly (zero-copy write)
        /// Returns a span where you can write your data directly
        /// Call CommitFrame() after writing to complete the operation, can throw ReaderDeadException, FrameTooLargeException, BufferFullException
        /// </summary>
        public unsafe Span<byte> GetFrameBuffer(int size, out ulong sequenceNumber)
        {
            ThrowIfDisposed();

            _logger.LogDebug("GetFrameBuffer called with size={Size}", size);

            // Check frame size
            long totalFrameSize = FrameHeader.SIZE + size;
            
            // Use ReadRef for initial checks to avoid copying
            ref readonly var oiebRef = ref _sharedMemory.ReadRef<OIEB>(0);
            if(oiebRef.ReaderPid == 0)
                throw new ReaderDeadException();

            if ((ulong)totalFrameSize > oiebRef.PayloadSize)
            {
                _logger.LogError("Frame too large: {FrameSize} > {PayloadSize}", totalFrameSize, oiebRef.PayloadSize);
                throw new FrameTooLargeException();
            }

            // Wait for space
            while (true)
            {
                // First check current state
                oiebRef = ref _sharedMemory.ReadRef<OIEB>(0);
                
                // Calculate required space including potential wrap-around waste
                long spaceToEndCheck = (long)oiebRef.PayloadSize - (long)oiebRef.PayloadWritePos;
                ulong requiredSpace = (spaceToEndCheck >= totalFrameSize)
                    ? (ulong)totalFrameSize                                      // no wrap needed
                    : (ulong)spaceToEndCheck + (ulong)totalFrameSize;            // wrap: waste at end + frame at beginning

                if (oiebRef.PayloadFreeBytes >= requiredSpace)
                    break;

                // Need to wait for space - wait on semaphore
                if (!_readSemaphore.Wait(TimeSpan.FromSeconds(5)))
                {
                    // Timeout - check if reader died
                    oiebRef = ref _sharedMemory.ReadRef<OIEB>(0);
                    if (oiebRef.ReaderPid == 0 || !ProcessExists(oiebRef.ReaderPid))
                    {
                        throw new ReaderDeadException();
                    }
                    throw new BufferFullException();
                }
                // Loop back to check space again after semaphore signal
            }
            
            // Get a reference for direct modifications
            ref var oieb = ref _sharedMemory.ReadRef<OIEB>(0);

            _logger.LogTrace("OIEB state before write: WrittenCount={WrittenCount}, ReadCount={ReadCount}, WritePos={WritePos}, ReadPos={ReadPos}, FreeBytes={FreeBytes}",
                oieb.PayloadWrittenCount, oieb.PayloadReadCount, oieb.PayloadWritePos, oieb.PayloadReadPos, oieb.PayloadFreeBytes);

            // Calculate write position
            long writePos = _payloadOffset + (long)oieb.PayloadWritePos;
            long spaceToEnd = _payloadOffset + (long)oieb.PayloadSize - writePos;

            // Check if we need to wrap
            if (totalFrameSize > spaceToEnd)
            {
                // Write wrap marker (protocol: payload_size = 0)
                var wrapHeader = new FrameHeader
                {
                    PayloadSize = 0,
                    SequenceNumber = 0
                };

                _sharedMemory.Write(writePos, wrapHeader);

                // Account for the wasted space at the end - atomic to prevent lost updates
                Interlocked.Add(ref Unsafe.As<ulong, long>(ref oieb.PayloadFreeBytes), -(long)spaceToEnd);

                // Update position to beginning
                oieb.PayloadWritePos = 0;
                oieb.PayloadWrittenCount++;
                writePos = _payloadOffset;

                // Check space again after wrap
                if ((ulong)Interlocked.Read(ref Unsafe.As<ulong, long>(ref oieb.PayloadFreeBytes)) < (ulong)totalFrameSize)
                {
                    throw new BufferFullException();
                }
            }

            // Write frame header
            sequenceNumber = _sequenceNumber++;
            var header = new FrameHeader
            {
                PayloadSize = (ulong)size,
                SequenceNumber = sequenceNumber
            };

            _sharedMemory.Write(writePos, header);

            // Get pointer to frame data area
            long dataPos = writePos + FrameHeader.SIZE;
            byte* framePtr = _sharedMemory.GetPointer(dataPos);

            // Store write position for commit
            _pendingWritePos = dataPos;
            _pendingFrameSize = size;

            return new Span<byte>(framePtr, size);
        }

        private long _pendingWritePos;
        private int _pendingFrameSize;

        /// <summary>
        /// Commit the frame after writing to the buffer returned by GetFrameBuffer
        /// </summary>
        public void CommitFrame()
        {
            ref var oieb = ref _sharedMemory.ReadRef<OIEB>(0);

            // Update write position directly in shared memory
            oieb.PayloadWritePos = (ulong)((_pendingWritePos + _pendingFrameSize - _payloadOffset) % (long)oieb.PayloadSize);
            oieb.PayloadWrittenCount++;

            // Update free bytes - atomic to prevent lost updates when reader adds concurrently
            long totalFrameSize = FrameHeader.SIZE + _pendingFrameSize;
            Interlocked.Add(ref Unsafe.As<ulong, long>(ref oieb.PayloadFreeBytes), -totalFrameSize);
            _sharedMemory.Flush();

            // Signal data available
            _writeSemaphore.Release();
        }

        /// <summary>
        /// Write a frame to the buffer (convenience method that copies data)
        /// For zero-copy writing, use GetFrameBuffer/CommitFrame instead
        /// </summary>
        public void WriteFrame(ReadOnlySpan<byte> data)
        {
            ThrowIfDisposed();

            // Check frame size
            long totalFrameSize = FrameHeader.SIZE + data.Length;
            ref readonly var oiebRef = ref _sharedMemory.ReadRef<OIEB>(0);

            if ((ulong)totalFrameSize > oiebRef.PayloadSize)
                throw new FrameTooLargeException();

            // Wait for space
            ref var oieb = ref _sharedMemory.ReadRef<OIEB>(0);
            while (true)
            {
                // Check if reader hasn't exited gracefully
                if (oieb.ReaderPid == 0)
                    throw new ReaderDeadException();
                
                // Calculate required space including potential wrap-around waste
                long spaceToEndCheck = (long)oieb.PayloadSize - (long)oieb.PayloadWritePos;
                ulong requiredSpace = (spaceToEndCheck >= totalFrameSize)
                    ? (ulong)totalFrameSize                                      // no wrap needed
                    : (ulong)spaceToEndCheck + (ulong)totalFrameSize;            // wrap: waste at end + frame at beginning

                if (oieb.PayloadFreeBytes >= requiredSpace)
                    break;

                // Need to wait for space - wait on semaphore
                if (_readSemaphore.Wait(_writeTimeout)) continue;

                if (oieb.ReaderPid == 0 || !ProcessExists(oieb.ReaderPid))
                {
                    throw new ReaderDeadException();
                }
                throw new BufferFullException();

            }

            // Calculate write position
            long writePos = _payloadOffset + (long)oieb.PayloadWritePos;
            long spaceToEnd = _payloadOffset + (long)oieb.PayloadSize - writePos;

            // Check if we need to wrap
            if (totalFrameSize > spaceToEnd)
            {
                // Write wrap marker (protocol: payload_size = 0)
                var wrapHeader = new FrameHeader
                {
                    PayloadSize = 0,
                    SequenceNumber = 0
                };

                _sharedMemory.Write(writePos, wrapHeader);

                // Account for the wasted space at the end - atomic to prevent lost updates
                Interlocked.Add(ref Unsafe.As<ulong, long>(ref oieb.PayloadFreeBytes), -(long)spaceToEnd);

                // Update position to beginning
                oieb.PayloadWritePos = 0;
                oieb.PayloadWrittenCount++;
                writePos = _payloadOffset;

                // Check space again after wrap
                if ((ulong)Interlocked.Read(ref Unsafe.As<ulong, long>(ref oieb.PayloadFreeBytes)) < (ulong)totalFrameSize)
                {
                    throw new BufferFullException();
                }
            }

            // Write frame header
            var header = new FrameHeader
            {
                PayloadSize = (ulong)data.Length,
                SequenceNumber = _sequenceNumber++
            };

            _sharedMemory.Write(writePos, header);

            // Write frame data
            long dataPos = writePos + FrameHeader.SIZE;
            _sharedMemory.WriteSpan(dataPos, data);

            // Now get a reference to update the OIEB in shared memory
            ref var oiebUpdate = ref _sharedMemory.ReadRef<OIEB>(0);

            // Update write position
            oiebUpdate.PayloadWritePos = (ulong)((dataPos + data.Length - _payloadOffset) % (long)oieb.PayloadSize);
            oiebUpdate.PayloadWrittenCount = oieb.PayloadWrittenCount + 1;

            // Update free bytes - atomic to prevent lost updates when reader adds concurrently
            Interlocked.Add(ref Unsafe.As<ulong, long>(ref oiebUpdate.PayloadFreeBytes), -(long)totalFrameSize);
            _sharedMemory.Flush();

            // Signal data available
            _writeSemaphore.Release();
        }

        /// <summary>
        /// Check if reader is connected
        /// </summary>
        public bool IsReaderConnected()
        {
            if (_disposed)
                return false;

            // Use ReadRef to avoid copying OIEB
            ref readonly var oieb = ref _sharedMemory.ReadRef<OIEB>(0);
            return oieb.ReaderPid != 0 && ProcessExists(oieb.ReaderPid);
        }

        public string Name => _name;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static long CalculateUsedBytes(ulong writePos, ulong readPos, ulong bufferSize)
        {
            return writePos >= readPos 
                ? (long)(writePos - readPos) 
                : (long)(bufferSize - readPos + writePos);
        }

        private static bool ProcessExists(ulong pid)
        {
            try
            {
                using var process = Process.GetProcessById((int)pid);
                return !process.HasExited;
            }
            catch
            {
                return false;
            }
        }

        private void ThrowIfDisposed()
        {
            ObjectDisposedException.ThrowIf(_disposed, this);
        }

        public void Dispose()
        {
            if (_disposed)
                return;

            _disposed = true;

            // Clear writer PID
            try
            {
                ref var oieb = ref _sharedMemory.ReadRef<OIEB>(0);
                oieb.WriterPid = 0;
                _sharedMemory.Flush();
            }
            catch { }

            _sharedMemory?.Dispose();
            _writeSemaphore?.Dispose();
            _readSemaphore?.Dispose();
        }
        
        /// <summary>
        /// Get the current OIEB state (for testing purposes)
        /// </summary>
        internal unsafe OIEB GetOIEB()
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(Writer));
                
            return _sharedMemory.ReadRef<OIEB>(0);
        }
    }
}